File: wick/term.py
"""
Core data structures for second-quantized operator algebra.

Operator input convention
--------------------------
Operators are specified in TRUE VACUUM notation:
  '+p'  -> a†_p  (creation operator)
  '-p'  -> a_p   (annihilation operator)

The Fermi vacuum type (dagger_fermi) is derived automatically from the label:
  a†_virt  ->  FV creator     (dagger_fermi = True)
  a†_occ   ->  FV annihilator (dagger_fermi = False)  -- fills a hole
  a_virt   ->  FV annihilator (dagger_fermi = False)
  a_occ    ->  FV creator     (dagger_fermi = True)   -- creates a hole

Index label conventions
-----------------------
  Occupied : labels starting with i, j, k, l, m, n
  Virtual  : labels starting with a, b, c, d, e, f
  General  : anything else (p, q, r, s, t, u, v, w, ...)
             General labels must be expanded before dagger_fermi is accessed.
"""

from __future__ import annotations
from typing import List, Optional

# ── index classification ────────────────────────────────────────────────────

_OCC_CHARS  = frozenset('ijklmn')
_VIRT_CHARS = frozenset('abcdef')


def is_occ(label: str) -> bool:
    # conventional: i,j,k,l,m,n...  OR  fresh expanded: o1, o2, ...
    return label[0] in _OCC_CHARS or (label[0] == 'o' and label[1:].isdigit())

def is_virt(label: str) -> bool:
    # conventional: a,b,c,d,e,f...  OR  fresh expanded: v1, v2, ...
    return label[0] in _VIRT_CHARS or (label[0] == 'v' and label[1:].isdigit())

def is_general(label: str) -> bool:
    return not is_occ(label) and not is_virt(label)


# ── Operator ────────────────────────────────────────────────────────────────

class Operator:
    """A single fermionic creation/annihilation operator (true vacuum basis)."""

    __slots__ = ('label', 'dagger')

    def __init__(self, label: str, dagger: bool):
        self.label  = label   # index label, e.g. 'p', 'i1', 'a'
        self.dagger = dagger  # True = a†_label, False = a_label (TRUE vacuum)

    @property
    def dagger_fermi(self) -> bool:
        """Creator relative to Fermi vacuum?

        dagger_fermi = dagger XOR is_occ(label)
        """
        if is_occ(self.label):
            return not self.dagger
        if is_virt(self.label):
            return self.dagger
        raise ValueError(
            f"dagger_fermi undefined for general label '{self.label}'. "
            "Expand general labels first."
        )

    def renamed(self, mapping: dict) -> 'Operator':
        return Operator(mapping.get(self.label, self.label), self.dagger)

    def copy(self) -> 'Operator':
        return Operator(self.label, self.dagger)

    def __repr__(self) -> str:
        return f"{'+'if self.dagger else '-'}{self.label}"

    def __eq__(self, other) -> bool:
        return isinstance(other, Operator) and self.label == other.label and self.dagger == other.dagger


# ── Tensor ──────────────────────────────────────────────────────────────────

class Tensor:
    """A named tensor with index labels."""

    __slots__ = ('name', 'indices')

    def __init__(self, name: str, indices: List[str]):
        self.name    = name
        self.indices = list(indices)

    def renamed(self, mapping: dict) -> 'Tensor':
        return Tensor(self.name, [mapping.get(i, i) for i in self.indices])

    def copy(self) -> 'Tensor':
        return Tensor(self.name, list(self.indices))

    def to_string(self) -> str:
        """Format as pdaggerq-style string."""
        if self.name == 'g':
            p, q, r, s = self.indices
            return f'<{p},{q}||{r},{s}>'
        if self.name == 'f':
            p, q = self.indices
            return f'f({p},{q})'
        return f'{self.name}({",".join(self.indices)})'

    def __repr__(self) -> str:
        return self.to_string()

    def __eq__(self, other) -> bool:
        return isinstance(other, Tensor) and self.name == other.name and self.indices == other.indices


# ── Delta ───────────────────────────────────────────────────────────────────

class Delta:
    """Kronecker delta  δ_{ij}."""

    __slots__ = ('i', 'j')

    def __init__(self, i: str, j: str):
        self.i = i
        self.j = j

    def renamed(self, mapping: dict) -> 'Delta':
        return Delta(mapping.get(self.i, self.i), mapping.get(self.j, self.j))

    def copy(self) -> 'Delta':
        return Delta(self.i, self.j)

    def __repr__(self) -> str:
        return f'd({self.i},{self.j})'

    def __eq__(self, other) -> bool:
        return isinstance(other, Delta) and self.i == other.i and self.j == other.j


# ── Term ────────────────────────────────────────────────────────────────────

class Term:
    """A single term: sign * factor * Π tensors * Π deltas * Π operators."""

    def __init__(
        self,
        factor:    float,
        sign:      int,
        tensors:   List[Tensor],
        operators: List[Operator],
        deltas:    Optional[List[Delta]] = None,
    ):
        self.factor    = factor
        self.sign      = sign          # +1 or -1
        self.tensors   = list(tensors)
        self.operators = list(operators)
        self.deltas    = list(deltas) if deltas else []

    @property
    def effective_factor(self) -> float:
        return self.sign * self.factor

    # ── bookkeeping ─────────────────────────────────────────────────────────

    def copy(self) -> 'Term':
        return Term(
            self.factor,
            self.sign,
            [t.copy() for t in self.tensors],
            [op.copy() for op in self.operators],
            [d.copy() for d in self.deltas],
        )

    def renamed(self, mapping: dict) -> 'Term':
        return Term(
            self.factor,
            self.sign,
            [t.renamed(mapping) for t in self.tensors],
            [op.renamed(mapping) for op in self.operators],
            [d.renamed(mapping) for d in self.deltas],
        )

    # ── normal-order queries ─────────────────────────────────────────────────

    def is_normal_ordered_fermi(self) -> bool:
        """True iff all FV creators precede all FV annihilators."""
        seen_annihilator = False
        for op in self.operators:
            df = op.dagger_fermi
            if not df:
                seen_annihilator = True
            elif seen_annihilator:
                return False
        return True

    def is_fully_contracted(self) -> bool:
        return len(self.operators) == 0

    def has_general_labels(self) -> bool:
        for op in self.operators:
            if is_general(op.label):
                return True
        for t in self.tensors:
            for idx in t.indices:
                if is_general(idx):
                    return True
        return False

    def __repr__(self) -> str:
        parts = [f'{self.effective_factor:+.6f}']
        parts += [t.to_string() for t in self.tensors]
        parts += [repr(d) for d in self.deltas]
        parts += [repr(op) for op in self.operators]
        return ' * '.join(parts)

File: src/term.rs
//! Core data structures for second-quantized operator algebra.
//!
//! Operator input convention (TRUE VACUUM):
//!   '+p'  -> a†_p  (creation operator)
//!   '-p'  -> a_p   (annihilation operator)
//!
//! Index label conventions:
//!   Occupied : i, j, k, l, m, n  OR  o1, o2, o3, ...
//!   Virtual  : a, b, c, d, e, f  OR  v1, v2, v3, ...
//!   General  : p, q, r, s, t, u, w  (must be expanded before use)

use std::collections::HashMap;
use std::fmt;

// ── index classification ────────────────────────────────────────────────────

const OCC_CHARS: &[char] = &['i', 'j', 'k', 'l', 'm', 'n'];
const VIRT_CHARS: &[char] = &['a', 'b', 'c', 'd', 'e', 'f'];

pub fn is_occ(label: &str) -> bool {
    let first = label.chars().next().unwrap();
    if OCC_CHARS.contains(&first) {
        return true;
    }
    first == 'o' && label.len() > 1 && label[1..].chars().all(|c| c.is_ascii_digit())
}

pub fn is_virt(label: &str) -> bool {
    let first = label.chars().next().unwrap();
    if VIRT_CHARS.contains(&first) {
        return true;
    }
    first == 'v' && label.len() > 1 && label[1..].chars().all(|c| c.is_ascii_digit())
}

pub fn is_general(label: &str) -> bool {
    !is_occ(label) && !is_virt(label)
}

// ── Operator ────────────────────────────────────────────────────────────────

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct Operator {
    pub label: String,
    pub dagger: bool,
}

impl Operator {
    pub fn new(label: &str, dagger: bool) -> Self {
        Self {
            label: label.to_string(),
            dagger,
        }
    }

    /// Creator relative to Fermi vacuum?
    /// dagger_fermi = dagger XOR is_occ(label)
    pub fn dagger_fermi(&self) -> bool {
        if is_occ(&self.label) {
            return !self.dagger;
        }
        if is_virt(&self.label) {
            return self.dagger;
        }
        panic!(
            "dagger_fermi undefined for general label '{}'. Expand general labels first.",
            self.label
        );
    }

    pub fn renamed(&self, mapping: &HashMap<String, String>) -> Self {
        Self {
            label: mapping.get(&self.label).cloned().unwrap_or_else(|| self.label.clone()),
            dagger: self.dagger,
        }
    }
}

impl fmt::Display for Operator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}{}", if self.dagger { "+" } else { "-" }, self.label)
    }
}

impl fmt::Debug for Operator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self, f)
    }
}

// ── Tensor ──────────────────────────────────────────────────────────────────

#[derive(Clone, PartialEq, Eq)]
pub struct Tensor {
    pub name: String,
    pub indices: Vec<String>,
}

impl Tensor {
    pub fn new(name: &str, indices: Vec<&str>) -> Self {
        Self {
            name: name.to_string(),
            indices: indices.into_iter().map(|s| s.to_string()).collect(),
        }
    }

    pub fn from_strings(name: String, indices: Vec<String>) -> Self {
        Self { name, indices }
    }

    pub fn renamed(&self, mapping: &HashMap<String, String>) -> Self {
        Self {
            name: self.name.clone(),
            indices: self
                .indices
                .iter()
                .map(|i| mapping.get(i).cloned().unwrap_or_else(|| i.clone()))
                .collect(),
        }
    }

    pub fn to_string_pdaggerq(&self) -> String {
        if self.name == "g" {
            format!(
                "<{},{}||{},{}>",
                self.indices[0], self.indices[1], self.indices[2], self.indices[3]
            )
        } else if self.name == "f" {
            format!("f({},{})", self.indices[0], self.indices[1])
        } else {
            format!(
                "{}({})",
                self.name,
                self.indices.join(",")
            )
        }
    }
}

impl fmt::Display for Tensor {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.to_string_pdaggerq())
    }
}

impl fmt::Debug for Tensor {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self, f)
    }
}

// ── Delta ───────────────────────────────────────────────────────────────────

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct Delta {
    pub i: String,
    pub j: String,
}

impl Delta {
    pub fn new(i: &str, j: &str) -> Self {
        Self {
            i: i.to_string(),
            j: j.to_string(),
        }
    }

    pub fn renamed(&self, mapping: &HashMap<String, String>) -> Self {
        Self {
            i: mapping.get(&self.i).cloned().unwrap_or_else(|| self.i.clone()),
            j: mapping.get(&self.j).cloned().unwrap_or_else(|| self.j.clone()),
        }
    }
}

impl fmt::Display for Delta {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "d({},{})", self.i, self.j)
    }
}

impl fmt::Debug for Delta {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self, f)
    }
}

// ── Term ────────────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct Term {
    pub factor: f64,
    pub sign: i32,
    pub tensors: Vec<Tensor>,
    pub operators: Vec<Operator>,
    pub deltas: Vec<Delta>,
}

impl Term {
    pub fn new(
        factor: f64,
        sign: i32,
        tensors: Vec<Tensor>,
        operators: Vec<Operator>,
        deltas: Vec<Delta>,
    ) -> Self {
        Self {
            factor,
            sign,
            tensors,
            operators,
            deltas,
        }
    }

    pub fn effective_factor(&self) -> f64 {
        self.sign as f64 * self.factor
    }

    pub fn renamed(&self, mapping: &HashMap<String, String>) -> Self {
        Self {
            factor: self.factor,
            sign: self.sign,
            tensors: self.tensors.iter().map(|t| t.renamed(mapping)).collect(),
            operators: self.operators.iter().map(|op| op.renamed(mapping)).collect(),
            deltas: self.deltas.iter().map(|d| d.renamed(mapping)).collect(),
        }
    }

    pub fn is_normal_ordered_fermi(&self) -> bool {
        let mut seen_annihilator = false;
        for op in &self.operators {
            let df = op.dagger_fermi();
            if !df {
                seen_annihilator = true;
            } else if seen_annihilator {
                return false;
            }
        }
        true
    }

    pub fn is_fully_contracted(&self) -> bool {
        self.operators.is_empty()
    }

    pub fn has_general_labels(&self) -> bool {
        for op in &self.operators {
            if is_general(&op.label) {
                return true;
            }
        }
        for t in &self.tensors {
            for idx in &t.indices {
                if is_general(idx) {
                    return true;
                }
            }
        }
        false
    }
}

impl fmt::Display for Term {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut parts = vec![format!("{:+.6}", self.effective_factor())];
        for t in &self.tensors {
            parts.push(t.to_string_pdaggerq());
        }
        for d in &self.deltas {
            parts.push(format!("{}", d));
        }
        for op in &self.operators {
            parts.push(format!("{}", op));
        }
        write!(f, "{}", parts.join(" * "))
    }
}

impl fmt::Debug for Term {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self, f)
    }
}

// ── Parsing helpers ─────────────────────────────────────────────────────────

pub fn parse_tensor(s: &str) -> Tensor {
    let s = s.trim();
    if let Some(paren_start) = s.find('(') {
        let name = &s[..paren_start];
        let rest = &s[paren_start + 1..s.len() - 1];
        let indices: Vec<String> = rest.split(',').map(|x| x.trim().to_string()).collect();
        Tensor::from_strings(name.to_string(), indices)
    } else {
        panic!("Cannot parse tensor string: '{}'", s);
    }
}

pub fn parse_op(s: &str) -> Operator {
    let s = s.trim();
    let first = s.chars().next().unwrap();
    match first {
        '+' => Operator::new(&s[1..], true),
        '-' => Operator::new(&s[1..], false),
        _ => panic!("Operator string must start with '+' or '-', got: '{}'", s),
    }
}
